File: TpDotNetCore/Entities/Punch.cs
using System;

namespace TpDotNetCore.Entities
{
    public class Punch
    {
        public string Id { get; set; }

        public DateTime Created { get; set; }
        public DateTime Updated { get; set; }
        
        public DateTime PunchTime { get; set; }
        public decimal TimeDec { get; set; }
        public bool Direction { get; set; } // true: In, false: Out

        public int DayPunchId { get; set; }
        public DayPunch DayPunch { get; set; }
        public int WeekPunchId { get; set; }
        public WeekPunch WeekPunch { get; set; }
        public int MonthPunchId { get; set; }
        public MonthPunch MonthPunch { get; set; }
        public int YearPunchId { get; set; }
        public YearPunch YearPunch { get; set; }
        public string UserId { get; set; }
        public User User { get; set; }
    }
}

File: TpServer/src/entity/punch.ts
import { Table, Column, CreateDateColumn, UpdateDateColumn, PrimaryGeneratedColumn, ManyToOne, VersionColumn } from "typeorm";
import { User } from './user';
import { YearPunch } from './yearpunch';
import { MonthPunch } from './monthpunch';
import { WeekPunch } from './weekpunch';
import { DayPunch } from './daypunch';

@Table()
export class Punch {
    @PrimaryGeneratedColumn()
    id: number;
    @Column("time")
    time: Date;
    @Column()
    timedec: number;
    @Column()
    direction: boolean; // true: In, false: Out
    @ManyToOne(type => User, user => user.punches)
    user: User;
    @ManyToOne(type => YearPunch, year => year.punches)
    year: YearPunch;
    @ManyToOne(type => MonthPunch, month => month.punches)
    month: MonthPunch;
    @ManyToOne(type => WeekPunch, week => week.punches)
    week: WeekPunch;
    @ManyToOne(type => DayPunch, day => day.punches)
    day: DayPunch;
    @CreateDateColumn()
    created: Date;
    @UpdateDateColumn()
    updated: Date;
    @VersionColumn()
    version: number;
}